book/src/proposals/abi-management.md

```diff
@@ -0,0 +1,434 @@
+# Solana ABI management process
+
+This document proposes Solana ABI management process. The ABI management process
+is an engineering practice and supporting technical framework to avoid
+introducing unintended incompatible ABI changes.
+
+# Problem
+
+The Solana ABI (binary interface to the cluster) is currently only defined 
+implicitly by the implementation, and requires a very careful eye to notice
+breaking changes. This makes it extremely difficult to upgrade the software
+on an existing cluster without rebooting the ledger.
+
+# Requirements and objectives
+
+- Unintended ABI changes can be detected as CI failures mechanically.
+- Newer impl must be able to process oldest data (since genesis) once we go
+  mainnet. (This is in stark contrast with regard to conventional software;
+  A decade-old SSL 3.0 doesn't work anymore, but the genesis block of Bitcoin
+  is still replayable today):
+  - Deserialization stability: Newer implementation (like system programs)
+    processes past object(transactions) exactly in the identical way when
+    replaying the ledger.
+  - Serialization stability: Newer implementation (like AccountsDB) produces
+    newer blobs exactly in the identical way unless its ABI is explicitly and
+    knowingly changed.
+- The objective of this proposal is to protect the ABI while sustaining rather
+  rapid development by opting for mechanical process rather than very long
+  human-driven auditing process. After all, while cluster instability must be
+  avoided, timely delivery of software improvements is desired for all
+  stakeholders of the Solana cluster.
```

Comment:
> After all, while cluster instability must be avoided, timely delivery of software improvements is desired for all stakeholders of the Solana cluster.

I think this is a given, no need to explicitly call it out :)

Reply:
Removed!